### ISSUE TITLE
Increase source count limit

### ISSUE THREAD

**Issue by Deer-Spangle** [zwagoth/e621ng#427]
As mildly discussed in https://e621.net/forum_topics/34730

Currently there's a limit on source links for posts; there can be no more than 10 per post. There are 501 posts which hit this limit, and it seems rather restrictive?
At a glance, 10 source links sounds like a lot! But given the recommendation to have 3 links per actual source (submission link, direct file link, and user gallery link), that's actually only really 3 actual sources that are allowed.
Uploading an image to twitter, Furaffinity and Weasyl would take all your source links, and throw in deviantart, or furry network or something, and you run out.

I get that there must be a limit to prevent abuse, and I feel like moving it to 30 would be more than enough, without being an undue burden.

There are 501 posts hitting the cap of 10 sources
But no posts come close to hitting the cap of 2048 character long sources. The closest is a google translate link in [post 53245](https://e621.net/posts/53245), which is 1655 characters long. (That source should probably be updated or removed.) After that one, the next ones are all around 1000 characters long, about half the limit.
Obviously, more rows is more of a database burden than more columns, but it seems like sources are stored newline-delimited in one column anway. Rather than as a separate table.
Given the source link length limit is so generous, it seems like the number of source links should also be a bit more generous than it is. Especially given the recommendations on source link usage carve that down to a third.

**Comment by Earlopain** [comment]
Sources are stored newline delimited in one column, yes. It's just a matter of increasing the number, I don't think performance is going to matter.

Personally I don't understand the need to add gallery urls. One direct link I can understand, from where the file was actually sourced/uploaded from. But the gallery links are available when going to the submission yourself, and the artist entry can contain a bunch of gallery links as well. Everything else just looks like clutter.

**Comment by DonovanDMC** [comment]
The wisdom I've gotten is artist url & source url, but only the former when the source url does not contain the artist url

So for twitter, that's one url, as the tweet link contains the artist's name in the link

For furaffinity, that's two, as the source url does not contain the artist's name

(quote from Mairo, in reference to asking if the base twitter url should be provided alongside the tweet url)

> In that instance, I deem it reduntant and more reduntant information in the source field, the harder it becomes to quickly parse it. Because even if the submission dies, the information of where it's from is still directly on that submission link (individual tweet). Additionally it normalizes that so many people simply link to artist galleries (in this case twitter account) when there's no other source and that's doesn't in itself count as source and is also again reduntant. 

> With websites where the submission or direct link doesn't directly correlate to the gallery where it's from (e.g. you cannot reverse into gallery where it's from even by taking specific URL parts of submission or direct link), then the gallery link is important, but only as addition and last on the list.


![Screenshot_20220806-150618.png](https://user-images.githubusercontent.com/17226394/183264553-652aa59b-c006-405e-ae04-afc1d86c1053.png)

**Comment by Earlopain** [comment]
Yeah, that makes sense, not something I would do myself though. Increasing it to 30 when only 500 posts are hitting the current limit seems a bit high. Considering that there are that few posts where this is a problem, users simply aren't sourcing so that this becomes an issue (and I am part of the problem). I do wonder what the overall source count distribution looks like, that would be interesting.

Posts like [this](https://e926.net/posts/1250892) might be problematic, but I think you would be hard-pressed to find a post with all of them as sources, since more than half of them aren't found by saucenao. With the rule by mairo you would land at 12 sources, or 13 when direct uploading.
![image](https://user-images.githubusercontent.com/14981592/183265795-bdf0d826-1bf0-43b5-8dae-3d6eab57a90d.png)

I'm not opposed against a small increase, to give the current posts at the limit a bit of breathing room. But my overall opinion of this is that it's trying to solve a problem that doesn't exist.

**Comment by Deer-Spangle** [comment]
Here's the counter by number of sources:
> 0 links: 383608
> 1 link: 1,305,003
> 2 links: 1,003,425
> 3 links: 566,133
> 4 links: 132,653
> 5 links: 59,281
> 6 links: 12,143
> 7 links: 4,529
> 8 links: 1,639
> 9 links: 843
> 10 links: 501

![image](https://user-images.githubusercontent.com/60626596/183269260-2d3e57cb-f8c7-4ed9-86a3-2f885207fe84.png)
And the log of that, for whatever's that worth
![image](https://user-images.githubusercontent.com/60626596/183269278-c61f1b64-822d-4667-a708-cc0471300e50.png)

My view is not "only 500 are hitting the limit" so much as "this limit should only exist to restrict abuse, and those 500 are not being abusive, so it is over-restrictive"

I've heard folks advocate gallery URLs to me lately, as being very useful when an artist nukes their gallery, or prunes old content. 
And yes, that's less applicable with twitter, but twitter links do not enforce the username is correct at all. They only parse the tweet ID. I've used services that only store the tweet ID, and give links of the form: https://twitter.com/user/status/1551569879148421124
Weasyl also does not enforce the username in links, nor does deviantart. (And FA doesn't include the username, as noted)
I personally, find the direct image links more useless in the source list, but those are often used by folks to upload the image, I suppose!
I do sometimes use e621 sources to see if artists have galleries on different sites, which can be handy, and hence advocating for more sources.

Perhaps 30 is overkill, and 15 or 20 would be a better number. 
And yeah, real easy change to make, hence I figured I would make a thread for it, while my source cleaning PR is open, and to keep the bikeshedding separate over here ^_^

**Comment by Earlopain** [comment]
Nice graph, but it's rather sad that so many posts only have more than 1 source.

**Comment by DonovanDMC** [comment]
>  but twitter links do not enforce the username is correct at all.

I still don't see this as a valid reason to introduce a source that's almost always not needed
The url can be easily reformatted into a version that does contain their username. And by the logic of it not being tested, how do we know that base url is correct? We don't, and we would have to go searching through their entire tweet history to find out. If people are using an incorrect username in the tweet url, they're almost certainly doing that on purpose and could be reported for that. And if twitter doesn't check that, it's an easy fix. Someone will notice it eventually, and it will get replaced.

And I completely agree with another thing Mairo has said.
> Additionally it normalizes that so many people simply link to artist galleries (in this case twitter account) when there's no other source and that's doesn't in itself count as source and is also again reduntant.

It absolutely does. There are way too many posts where people just toss in a link to the artists twitter or fa gallery and call that a valid source. Doubling up sources encourages people to be lazy.

**Comment by Deer-Spangle** [comment]
[The upload form ](https://e621.net/uploads/new)as it stands recommends adding a gallery link alongside a submission link:
> You should include: A link to the artists page where this was obtained, and a link to the submission page where this image was obtained. No available source should ONLY be used if the content has never been posted online anywhere else. 

And yes, I also find the frequency of "gallery without submission link" frustrating. That was the original reason I built [the e621 source checker ](https://github.com/Deer-Spangle/e621_source_cleanup)which spurred [my thread on the forums](https://e621.net/forum_topics/34730).

There are 100k FA submissions with this, and while it is frustrating, it is much more of a solvable problem with the user gallery, than without. 
I do plan to go through and fix those to have a submission link, where possible.


Worth noting, that the code as it stands does support automatically adding gallery and direct links from user-given source links (though adding gallery links is the only one implemented) and then truncates the list down to 10 after doing:
```
sources + submission_sources + gallery_sources + direct_sources + additional_sources
).compact.reject{ |e| e.strip.empty? }.uniq
```
Meaning that it values, in order:
1. Source links given by the user or already there
2. Submission links
3. Gallery links
4. Direct image links
5. Additional links

So if the source list goes over 10, it cuts off additional links, then direct image links, then gallery links, etc.
But as I said, the handlers we have only actually generate gallery links at the moment.


And yeah, it is sad to see so many with only 1 source! I would like to fix that

**Comment by Deer-Spangle** [comment]
> There are way too many posts where people just toss in a link to the artists twitter or fa gallery and call that a valid source.

Started fixing this by the way. Threw some second project in [the repo I linked above](https://github.com/Deer-Spangle/e621_source_cleanup/tree/master/e621_gallery_finder).
I'm manually checking the matches are correct, and going one by one at the moment, but will optimise over time.
Current status is that it found 12155 direct submission links where the perceptual hashes and the gallery name matches the gallery specified in the sources. I've gone through a couple hundred and they're all good.

There are a few that match the perceptual hashes I have on file, but have been removed from FA since those hashes were generated. I'm not sure whether to make the e621 sources link to those deleted submissions in those cases.
